In [4]:
import sys

print(sys.executable)

c:\Users\DELL\OneDrive\Desktop\EffiSense-Keyboard-Productivity\venv\Scripts\python.exe


In [5]:
import pandas as pd

data_path = "../data/raw/free-text.csv"

df = pd.read_csv(data_path)

print("Shape:", df.shape)
print("\nColumns:")
print(df.columns.tolist())

Shape: (562583, 10)

Columns:
['participant', 'session', 'key1', 'key2', 'DU.key1.key1', 'DD.key1.key2', 'DU.key1.key2', 'UD.key1.key2', 'UU.key1.key2 ', 'Unnamed: 9']


C:\Users\DELL\AppData\Local\Temp\ipykernel_23820\1087466777.py:5: DtypeWarning: Columns (4: DU.key1.key1) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv(data_path)


In [6]:
print("Dataset shape:", df.shape)

print("\n--- Data types ---")
print(df.dtypes)

print("\n--- Missing values ---")
print(df.isnull().sum())

print("\n--- Duplicate rows ---")
print(df.duplicated().sum())

Dataset shape: (562583, 10)

--- Data types ---
participant          str
session            int64
key1                 str
key2                 str
DU.key1.key1      object
DD.key1.key2     float64
DU.key1.key2     float64
UD.key1.key2     float64
UU.key1.key2     float64
Unnamed: 9       float64
dtype: object

--- Missing values ---
participant           0
session               0
key1                  0
key2                198
DU.key1.key1          0
DD.key1.key2        198
DU.key1.key2        198
UD.key1.key2        198
UU.key1.key2        198
Unnamed: 9       562570
dtype: int64

--- Duplicate rows ---
7385


In [7]:
print("Number of participants:", df["participant"].nunique())
print("Number of sessions:", df["session"].nunique())

print("\nParticipants:")
print(df["participant"].unique())

print("\nEvents per participant:")
print(df.groupby("participant").size())

print("\nSessions per participant:")
print(df.groupby("participant")["session"].nunique())

Number of participants: 99
Number of sessions: 2

Participants:
<StringArray>
['p001', 'p002', 'p003', 'p004', 'p005', 'p006', 'p007', 'p008', 'p009',
 'p010', 'p011', 'p012', 'p013', 'p014', 'p015', 'p016', 'p017', 'p018',
 'p019', 'p020', 'p021', 'p022', 'p023', 'p024', 'p025', 'p026', 'p027',
 'p028', 'p029', 'p030', 'p031', 'p032', 'p033', 'p034', 'p035', 'p036',
 'p037', 'p038', 'p039', 'p040', 'p041', 'p042', 'p043', 'p044', 'p045',
 'p046', 'p047', 'p048', 'p049', 'p050', 'p051', 'p052', 'p053', 'p054',
 'p055', 'p056', 'p057', 'p058', 'p059', 'p060', 'p061', 'p062', 'p063',
 'p064', 'p065', 'p066', 'p067', 'p068', 'p069', 'p070', 'p071', 'p072',
 'p073', 'p074', 'p075', 'p076', 'p077', 'p078', 'p079', 'p080', 'p081',
 'p082', 'p083', 'p084', 'p085', 'p086', 'p087', 'p088', 'p089', 'p090',
 'p091', 'p092', 'p093', 'p094', 'p095', 'p097', 'p098', 'p099', 'p100']
Length: 99, dtype: str

Events per participant:
participant
p001    5258
p002    5517
p003    5311
p004    4738
p005   

In [8]:
print("Session values:")
print(df["session"].value_counts().sort_index())

print("\nParticipant-session combinations:")
print(df.groupby(["participant", "session"]).size().describe())

print("\nNumber of participant-session groups:",
      df.groupby(["participant", "session"]).ngroups)

Session values:
session
1    278186
2    284397
Name: count, dtype: int64

Participant-session combinations:
count     198.000000
mean     2841.328283
std       316.532127
min        38.000000
25%      2702.750000
50%      2820.500000
75%      2947.750000
max      4700.000000
dtype: float64

Number of participant-session groups: 198


In [9]:
timing_cols = [
    "DU.key1.key1",
    "DD.key1.key2",
    "DU.key1.key2",
    "UD.key1.key2",
    "UU.key1.key2 "
]

for col in timing_cols:
    print(f"\n===== {col} =====")
    print("Data type:", df[col].dtype)
    print("Unique values:", df[col].nunique())
    print("Sample values:")
    print(df[col].head(10).tolist())


===== DU.key1.key1 =====
Data type: object
Unique values: 4501
Sample values:
['0.15', '0.962', '0.107', '0.08', '0.08', '0.16', '0.104', '0.073', '0.164', '0.063']

===== DD.key1.key2 =====
Data type: float64
Unique values: 8988
Sample values:
[-0.796, 1.148, 0.172, 0.2, 0.32, 1.432, 0.28, 0.196, 0.813, 0.183]

===== DU.key1.key2 =====
Data type: float64
Unique values: 5758
Sample values:
[0.166, 1.255, 0.252, 0.28, 0.48, 1.536, 0.353, 0.36, 0.876, 0.311]

===== UD.key1.key2 =====
Data type: float64
Unique values: 6935
Sample values:
[-0.946, 0.186, 0.065, 0.12, 0.24, 1.272, 0.176, 0.123, 0.649, 0.12]

===== UU.key1.key2  =====
Data type: float64
Unique values: 3417
Sample values:
[0.016, 0.293, 0.145, 0.2, 0.4, 1.376, 0.249, 0.287, 0.712, 0.248]


In [10]:
print("Unnamed: 9 non-null values:", df["Unnamed: 9"].notna().sum())
print("Unnamed: 9 unique values:", df["Unnamed: 9"].dropna().unique()[:20])

Unnamed: 9 non-null values: 13
Unnamed: 9 unique values: [0.185 0.042 0.464 0.072 0.28  0.345 0.065 0.331 0.064 0.12  0.192 0.254
 0.328]


In [11]:
extra_rows = df[df["Unnamed: 9"].notna()]

print("Number of rows:", len(extra_rows))

display(extra_rows)

Number of rows: 13


,participant,session,key1,key2,DU.key1.key1,DD.key1.key2,DU.key1.key2,UD.key1.key2,UU.key1.key2,Unnamed: 9
23481,p005,1,2,",0.087,1.088,1.183,1.001,1.096,\np005,1,",Shift,0.095,-0.472,0.280,-0.567,0.185
169381,p031,1,Backspace,",0.081,0.761,0.896,0.68,0.815,\np031,1,",Shift,0.135,-0.341,0.177,-0.476,0.042
177980,p033,1,e,",0.088,0.2,0.28,0.112,0.192\np033,1,",Backspace,0.080,0.400,0.544,0.320,0.464
181842,p033,2,t,",0.079,0.184,0.263,0.105,0.184,\np033,2,",Shift,0.079,-0.041,0.151,-0.120,0.072
204716,p037,2,Backspace,",0.134,0.281,0.502,0.147,0.368,\np037,2,",Backspace,0.221,0.342,0.501,0.121,0.280
205245,p037,2,Backspace,",0.166,0.569,0.726,0.403,0.56,\np037,2,",Backspace,0.157,0.319,0.502,0.162,0.345
306776,p055,1,Backspace,",0.123,0.785,0.87,0.662,0.747,\np055,1,",Shift,0.085,-0.168,0.150,-0.253,0.065
319078,p057,1,y,",0.137,0.304,0.49,0.167,0.353,\np057,1,",Shift,0.186,-0.047,0.517,-0.233,0.331
491603,p087,1,Backspace,",0.096,6.082,6.177,5.986,6.081,\np087,1,",Shift,0.095,-0.289,0.159,-0.384,0.064
491610,p087,1,Shift,",0.472,1.76,1.888,1.288,1.416,\np087,1,",Shift,0.128,-0.352,0.248,-0.480,0.120


In [12]:
df.loc[df["Unnamed: 9"].notna()]

,participant,session,key1,key2,DU.key1.key1,DD.key1.key2,DU.key1.key2,UD.key1.key2,UU.key1.key2,Unnamed: 9
23481,p005,1,2,",0.087,1.088,1.183,1.001,1.096,\np005,1,",Shift,0.095,-0.472,0.280,-0.567,0.185
169381,p031,1,Backspace,",0.081,0.761,0.896,0.68,0.815,\np031,1,",Shift,0.135,-0.341,0.177,-0.476,0.042
177980,p033,1,e,",0.088,0.2,0.28,0.112,0.192\np033,1,",Backspace,0.080,0.400,0.544,0.320,0.464
181842,p033,2,t,",0.079,0.184,0.263,0.105,0.184,\np033,2,",Shift,0.079,-0.041,0.151,-0.120,0.072
204716,p037,2,Backspace,",0.134,0.281,0.502,0.147,0.368,\np037,2,",Backspace,0.221,0.342,0.501,0.121,0.280
205245,p037,2,Backspace,",0.166,0.569,0.726,0.403,0.56,\np037,2,",Backspace,0.157,0.319,0.502,0.162,0.345
306776,p055,1,Backspace,",0.123,0.785,0.87,0.662,0.747,\np055,1,",Shift,0.085,-0.168,0.150,-0.253,0.065
319078,p057,1,y,",0.137,0.304,0.49,0.167,0.353,\np057,1,",Shift,0.186,-0.047,0.517,-0.233,0.331
491603,p087,1,Backspace,",0.096,6.082,6.177,5.986,6.081,\np087,1,",Shift,0.095,-0.289,0.159,-0.384,0.064
491610,p087,1,Shift,",0.472,1.76,1.888,1.288,1.416,\np087,1,",Shift,0.128,-0.352,0.248,-0.480,0.120


In [13]:
with open("../data/raw/free-text.csv", "r", encoding="utf-8") as f:
    lines = f.readlines()

print(lines[23481])
print(lines[23482])

p005,1,2,2,0.103,0.192,0.279,0.089,0.176,

p005,1,2,",0.087,1.088,1.183,1.001,1.096,



In [14]:
print("Rows with missing key2:", df["key2"].isna().sum())

print("Rows with missing timing values:")

print(df[
    [
        "DU.key1.key1",
        "DD.key1.key2",
        "DU.key1.key2",
        "UD.key1.key2",
        "UU.key1.key2 "
    ]
].isna().sum())

Rows with missing key2: 198
Rows with missing timing values:
DU.key1.key1       0
DD.key1.key2     198
DU.key1.key2     198
UD.key1.key2     198
UU.key1.key2     198
dtype: int64


In [15]:
print(df.columns.tolist())

['participant', 'session', 'key1', 'key2', 'DU.key1.key1', 'DD.key1.key2', 'DU.key1.key2', 'UD.key1.key2', 'UU.key1.key2 ', 'Unnamed: 9']


In [16]:
missing_key2 = df[df["key2"].isna()]

print("Number of rows:", len(missing_key2))
display(missing_key2.head(10))

Number of rows: 198


,participant,session,key1,key2,DU.key1.key1,DD.key1.key2,DU.key1.key2,UD.key1.key2,UU.key1.key2,Unnamed: 9
2513,p001,1,e,NaN,0.136,NaN,NaN,NaN,NaN,NaN
5257,p001,2,s,NaN,0.143,NaN,NaN,NaN,NaN,NaN
7959,p002,1,e,NaN,0.1,NaN,NaN,NaN,NaN,NaN
10774,p002,2,s,NaN,0.074,NaN,NaN,NaN,NaN,NaN
13367,p003,1,e,NaN,0.076,NaN,NaN,NaN,NaN,NaN
16085,p003,2,r,NaN,0.106,NaN,NaN,NaN,NaN,NaN
20785,p004,1,.,NaN,8.647,NaN,NaN,NaN,NaN,NaN
20823,p004,2,Space,NaN,178.759,NaN,NaN,NaN,NaN,NaN
24119,p005,1,e,NaN,0.112,NaN,NaN,NaN,NaN,NaN
27529,p005,2,s,NaN,0.104,NaN,NaN,NaN,NaN,NaN


In [17]:
clean_df = df.dropna(
    subset=[
        "key2",
        "DD.key1.key2",
        "DU.key1.key2",
        "UD.key1.key2",
        "UU.key1.key2 "
    ]
).copy()

# Remove the malformed extra column
clean_df = clean_df.drop(columns=["Unnamed: 9"])

print("Original shape:", df.shape)
print("Cleaned shape:", clean_df.shape)
print("Rows removed:", len(df) - len(clean_df))

Original shape: (562583, 10)
Cleaned shape: (562385, 9)
Rows removed: 198


In [18]:
print("Missing values after cleaning:")
print(clean_df.isna().sum())

Missing values after cleaning:
participant      0
session          0
key1             0
key2             0
DU.key1.key1     0
DD.key1.key2     0
DU.key1.key2     0
UD.key1.key2     0
UU.key1.key2     0
dtype: int64


In [19]:
duplicates = clean_df[clean_df.duplicated(keep=False)]

print("Total duplicate rows:", len(duplicates))

display(duplicates.head(20))

Total duplicate rows: 10544


,participant,session,key1,key2,DU.key1.key1,DD.key1.key2,DU.key1.key2,UD.key1.key2,UU.key1.key2
2948,p001,2,o,u,0.152,0.088,0.200,-0.064,0.048
3392,p001,2,o,u,0.137,0.080,0.184,-0.057,0.047
4520,p001,2,o,u,0.137,0.080,0.184,-0.057,0.047
5065,p001,2,o,u,0.152,0.088,0.200,-0.064,0.048
5270,p002,1,Space,t,0.07,0.090,0.161,0.020,0.091
5691,p002,1,Backspace,Backspace,0.061,0.131,0.196,0.070,0.135
5885,p002,1,s,Space,0.096,0.111,0.176,0.015,0.080
6300,p002,1,o,Space,0.07,0.085,0.145,0.015,0.075
6927,p002,1,Backspace,Backspace,0.061,0.131,0.196,0.070,0.135
7043,p002,1,s,Space,0.096,0.111,0.176,0.015,0.080


In [20]:
print("Total rows:", len(clean_df))
print("Duplicate rows:", clean_df.duplicated().sum())
print("Rows after removing exact duplicates:", len(clean_df.drop_duplicates()))

Total rows: 562385
Duplicate rows: 7385
Rows after removing exact duplicates: 555000


In [21]:
print("Number of participants:", clean_df["participant"].nunique())

print("\nParticipants:")
print(sorted(clean_df["participant"].unique()))

print("\nSessions:")
print(sorted(clean_df["session"].unique()))

print("\nRows per participant-session:")
print(
    clean_df.groupby(["participant", "session"])
    .size()
    .describe()
)

Number of participants: 99

Participants:
['p001', 'p002', 'p003', 'p004', 'p005', 'p006', 'p007', 'p008', 'p009', 'p010', 'p011', 'p012', 'p013', 'p014', 'p015', 'p016', 'p017', 'p018', 'p019', 'p020', 'p021', 'p022', 'p023', 'p024', 'p025', 'p026', 'p027', 'p028', 'p029', 'p030', 'p031', 'p032', 'p033', 'p034', 'p035', 'p036', 'p037', 'p038', 'p039', 'p040', 'p041', 'p042', 'p043', 'p044', 'p045', 'p046', 'p047', 'p048', 'p049', 'p050', 'p051', 'p052', 'p053', 'p054', 'p055', 'p056', 'p057', 'p058', 'p059', 'p060', 'p061', 'p062', 'p063', 'p064', 'p065', 'p066', 'p067', 'p068', 'p069', 'p070', 'p071', 'p072', 'p073', 'p074', 'p075', 'p076', 'p077', 'p078', 'p079', 'p080', 'p081', 'p082', 'p083', 'p084', 'p085', 'p086', 'p087', 'p088', 'p089', 'p090', 'p091', 'p092', 'p093', 'p094', 'p095', 'p097', 'p098', 'p099', 'p100']

Sessions:
[np.int64(1), np.int64(2)]

Rows per participant-session:
count     198.000000
mean     2840.328283
std       316.532127
min        37.000000
25%      270

In [22]:
timing_cols = [
    "DU.key1.key1",
    "DD.key1.key2",
    "DU.key1.key2",
    "UD.key1.key2",
    "UU.key1.key2 "
]

print(clean_df[timing_cols].describe())

        DD.key1.key2   DU.key1.key2   UD.key1.key2  UU.key1.key2 
count  562385.000000  562385.000000  562385.000000  562385.000000
mean        0.291818       0.495465       0.088476       0.292115
std         5.968022       5.490270       5.479273       4.635057
min      -568.083000      -7.935000    -568.162000     -10.782000
25%         0.120000       0.216000       0.019000       0.118000
50%         0.174000       0.272000       0.081000       0.176000
75%         0.272000       0.379000       0.173000       0.272000
max      1712.727000    1713.061000    1712.648000    1712.982000


In [23]:
print("Smallest DD values:")
display(
    clean_df.nsmallest(10, "DD.key1.key2")[
        ["participant", "session", "key1", "key2",
         "DU.key1.key1", "DD.key1.key2",
         "DU.key1.key2", "UD.key1.key2", "UU.key1.key2 "]
    ]
)

print("Largest DD values:")
display(
    clean_df.nlargest(10, "DD.key1.key2")[
        ["participant", "session", "key1", "key2",
         "DU.key1.key1", "DD.key1.key2",
         "DU.key1.key2", "UD.key1.key2", "UU.key1.key2 "]
    ]
)

Smallest DD values:


,participant,session,key1,key2,DU.key1.key1,DD.key1.key2,DU.key1.key2,UD.key1.key2,UU.key1.key2
541906,p097,1,n,Meta,0.079,-568.083,0.487,-568.162,0.408
406648,p072,1,Backspace,.,0.13,-475.140,0.870,-475.270,0.740
543417,p097,2,l,Meta,0.076,-315.899,0.404,-315.975,0.328
503397,p089,1,t,',0.123,-281.285,0.927,-281.408,0.804
354487,p063,1,Shift,.,0.0,-264.820,0.123,-264.820,0.123
549046,p098,2,Backspace,',0.055,-259.656,0.243,-259.711,0.188
439006,p078,1,a,c,0.044,-253.217,0.074,-253.261,0.030
409122,p072,2,u,',0.099,-244.892,1.380,-244.991,1.281
354308,p063,1,Shift,.,0.0,-225.071,0.131,-225.071,0.131
409110,p072,2,Backspace,.,0.135,-221.903,0.468,-222.038,0.333


Largest DD values:


,participant,session,key1,key2,DU.key1.key1,DD.key1.key2,DU.key1.key2,UD.key1.key2,UU.key1.key2
281255,p050,2,Space,Shift,0.079,1712.727,1713.061,1712.648,1712.982
279007,p050,1,n,Space,0.068,1482.061,1482.148,1481.993,1482.080
224298,p041,1,Backspace,CapsLock,0.1,1292.476,1292.591,1292.376,1292.491
79620,p015,1,Space,T,0.088,1067.658,1067.746,1067.570,1067.658
213589,p039,1,Shift,Shift,0.327,780.867,781.129,780.540,780.802
499175,p088,2,Space,s,0.095,696.208,696.319,696.113,696.224
490969,p087,1,Space,Shift,0.095,671.597,671.848,671.502,671.753
428090,p076,1,s,Space,0.087,657.177,657.289,657.090,657.202
406649,p072,1,.,Space,476.01,476.031,476.154,0.021,0.144
404000,p071,2,",",Space,0.096,393.232,393.391,393.136,393.295


In [24]:
print(
    clean_df.groupby("participant")[[
        "DD.key1.key2",
        "DU.key1.key2",
        "UD.key1.key2",
        "UU.key1.key2 "
    ]].agg(["min", "max"])
    .sort_values(("DD.key1.key2", "min"))
    .head(15)
)

            DD.key1.key2          DU.key1.key2          UD.key1.key2           \
                     min      max          min      max          min      max   
participant                                                                     
p097            -568.083  202.461        0.030  569.068     -568.162    5.601   
p072            -475.140  476.031        0.020  476.154     -475.270  386.719   
p089            -281.285  282.202        0.028  282.356     -281.408   55.314   
p063            -264.820  264.852        0.015  265.008     -264.820   79.408   
p098            -259.656  260.757        0.037  260.796     -259.711  193.772   
p078            -253.217  253.247        0.030  253.321     -253.261  241.878   
p053            -215.977  216.801        0.021  216.866     -216.049   61.712   
p039            -215.568  780.867        0.016  781.129     -215.647  780.540   
p031            -188.859  351.269       -0.341  351.575     -188.967  351.189   
p076            -161.803  65

In [25]:
import os

print(os.listdir("../data/raw"))

['free-text.csv']


In [26]:
import pandas as pd

raw = pd.read_csv(
    "../data/raw/free-text.csv",
    dtype=str,
    keep_default_na=False
)

print(raw.head(15).to_string())

   participant session   key1   key2 DU.key1.key1 DD.key1.key2 DU.key1.key2 UD.key1.key2 UU.key1.key2  Unnamed: 9
0         p001       1      W  Shift         0.15       -0.796        0.166       -0.946         0.016           
1         p001       1  Shift      e        0.962        1.148        1.255        0.186         0.293           
2         p001       1      e  Space        0.107        0.172        0.252        0.065         0.145           
3         p001       1  Space      b         0.08          0.2         0.28         0.12           0.2           
4         p001       1      b      e         0.08         0.32         0.48         0.24           0.4           
5         p001       1      e      l         0.16        1.432        1.536        1.272         1.376           
6         p001       1      l      i        0.104         0.28        0.353        0.176         0.249           
7         p001       1      i      e        0.073        0.196         0.36        0.123

In [27]:
raw.columns = raw.columns.str.strip()

print(raw.columns.tolist())

['participant', 'session', 'key1', 'key2', 'DU.key1.key1', 'DD.key1.key2', 'DU.key1.key2', 'UD.key1.key2', 'UU.key1.key2', 'Unnamed: 9']


In [28]:
timing_cols = [
    "DU.key1.key1",
    "DD.key1.key2",
    "DU.key1.key2",
    "UD.key1.key2",
    "UU.key1.key2"
]

for col in timing_cols:
    raw[col] = pd.to_numeric(raw[col], errors="coerce")

print("DD - DU:")
print((raw["DD.key1.key2"] - raw["DU.key1.key2"]).describe())

print("\nUD - DD:")
print((raw["UD.key1.key2"] - raw["DD.key1.key2"]).describe())

print("\nDU_pair - DU_key1:")
print((raw["DU.key1.key2"] - raw["DU.key1.key1"]).describe())

DD - DU:
count    562385.000000
mean         -0.203647
std           2.945276
min        -568.570000
25%          -0.117000
50%          -0.091000
75%          -0.072000
max           0.567000
dtype: float64

UD - DD:
count    562385.000000
mean         -0.203342
std           2.935615
min        -568.570000
25%          -0.117000
50%          -0.091000
75%          -0.072000
max           0.464000
dtype: float64

DU_pair - DU_key1:
count    562372.000000
mean          0.292126
std           4.635109
min         -10.782000
25%           0.118000
50%           0.176000
75%           0.272000
max        1712.982000
dtype: float64


In [29]:
# Rows with the 10 smallest DD values
print("10 smallest DD values:\n")
print(
    raw.nsmallest(10, "DD.key1.key2")[
        [
            "participant",
            "session",
            "key1",
            "key2",
            "DU.key1.key1",
            "DD.key1.key2",
            "DU.key1.key2",
            "UD.key1.key2",
            "UU.key1.key2"
        ]
    ].to_string(index=False)
)

print("\n\n10 largest DD values:\n")
print(
    raw.nlargest(10, "DD.key1.key2")[
        [
            "participant",
            "session",
            "key1",
            "key2",
            "DU.key1.key1",
            "DD.key1.key2",
            "DU.key1.key2",
            "UD.key1.key2",
            "UU.key1.key2"
        ]
    ].to_string(index=False)
)

10 smallest DD values:

participant session      key1 key2  DU.key1.key1  DD.key1.key2  DU.key1.key2  UD.key1.key2  UU.key1.key2
       p097       1         n Meta         0.079      -568.083         0.487      -568.162         0.408
       p072       1 Backspace    .         0.130      -475.140         0.870      -475.270         0.740
       p097       2         l Meta         0.076      -315.899         0.404      -315.975         0.328
       p089       1         t    '         0.123      -281.285         0.927      -281.408         0.804
       p063       1     Shift    .         0.000      -264.820         0.123      -264.820         0.123
       p098       2 Backspace    '         0.055      -259.656         0.243      -259.711         0.188
       p078       1         a    c         0.044      -253.217         0.074      -253.261         0.030
       p072       2         u    '         0.099      -244.892         1.380      -244.991         1.281
       p063       1     Shift  

In [30]:
# Check different ranges of DD.key1.key2

dd = raw["DD.key1.key2"]

print("Total rows:", len(dd))
print("Negative values:", (dd < 0).sum())
print("Zero values:", (dd == 0).sum())
print("0 to 1:", ((dd >= 0) & (dd <= 1)).sum())
print("1 to 5:", ((dd > 1) & (dd <= 5)).sum())
print("5 to 10:", ((dd > 5) & (dd <= 10)).sum())
print("10 to 60:", ((dd > 10) & (dd <= 60)).sum())
print("Above 60:", (dd > 60).sum())
print("Above 300:", (dd > 300).sum())
print("Above 1000:", (dd > 1000).sum())

Total rows: 562583
Negative values: 8526
Zero values: 20
0 to 1: 538342
1 to 5: 13078
5 to 10: 865
10 to 60: 1296
Above 60: 278
Above 300: 18
Above 1000: 4


## Day 1 — Dataset Understanding

### Raw dataset
- Rows: 562,583
- Columns: 10
- Participants: 99
- Sessions: 2
- Participant-session groups: 198

### Cleaning observations
- 198 rows have missing `key2` and associated key-pair timing values.
- `Unnamed: 9` contains 13 non-null values and is associated with malformed rows.
- The 198 incomplete rows are excluded from the working dataset.
- Duplicate complete rows: 7,385.
- Duplicate rows are retained because repeated keyboard transitions may represent legitimate behavior.

### Timing observations
- Most `DD.key1.key2` values (538,342 rows) lie between 0 and 1.
- 8,526 values are negative.
- 278 values are above 60.
- 18 values are above 300.
- 4 values are above 1000.
- Extreme timing values are retained for now because their semantics have not yet been established from source metadata.

### Current working dataset
- Rows: 562,385
- Columns: 9
- Missing values: 0

In [31]:
group_sizes = (
    clean_df
    .groupby(["participant", "session"])
    .size()
    .reset_index(name="rows")
)

print(group_sizes.head(10))

print("\nNumber of participant-session groups:", len(group_sizes))

print("\nGroup size statistics:")
print(group_sizes["rows"].describe())

  participant  session  rows
0        p001        1  2513
1        p001        2  2743
2        p002        1  2701
3        p002        2  2814
4        p003        1  2592
5        p003        2  2717
6        p004        1  4699
7        p004        2    37
8        p005        1  3295
9        p005        2  3409

Number of participant-session groups: 198

Group size statistics:
count     198.000000
mean     2840.328283
std       316.532127
min        37.000000
25%      2701.750000
50%      2819.500000
75%      2946.750000
max      4699.000000
Name: rows, dtype: float64


In [32]:
small_groups = group_sizes[group_sizes["rows"] < 500]

print("Groups with fewer than 500 rows:")
print(small_groups.to_string(index=False))

print("\nNumber of small groups:", len(small_groups))

Groups with fewer than 500 rows:
participant  session  rows
       p004        2    37

Number of small groups: 1


In [33]:
print("Total rows in clean dataset:", len(clean_df))

print("Rows needed per 10,000 windows if distributed evenly:",
      len(clean_df) / 10000)

Total rows in clean dataset: 562385
Rows needed per 10,000 windows if distributed evenly: 56.2385


In [34]:
participant_sizes = (
    clean_df
    .groupby("participant")
    .size()
    .sort_values()
)

print("\nParticipant row statistics:")
print(participant_sizes.describe())


Participant row statistics:
count      99.000000
mean     5680.656566
std       385.411023
min      4736.000000
25%      5435.000000
50%      5653.000000
75%      5941.500000
max      6811.000000
dtype: float64


In [35]:
window_sizes = [25, 50, 75, 100, 150]

print("Estimated number of non-overlapping windows:\n")

for w in window_sizes:
    total_windows = (
        clean_df
        .groupby(["participant", "session"])
        .size()
        .floordiv(w)
        .sum()
    )
    
    print(f"Window size {w:3d} events -> {total_windows:,} windows")

Estimated number of non-overlapping windows:

Window size  25 events -> 22,395 windows
Window size  50 events -> 11,149 windows
Window size  75 events -> 7,398 windows
Window size 100 events -> 5,530 windows
Window size 150 events -> 3,650 windows


In [36]:
w = 50

window_counts = (
    clean_df
    .groupby(["participant", "session"])
    .size()
    .floordiv(w)
    .reset_index(name="windows")
)

print(window_counts["windows"].describe())

print("\nSmallest 10 groups:")
print(
    window_counts
    .sort_values("windows")
    .head(10)
    .to_string(index=False)
)

print("\nLargest 10 groups:")
print(
    window_counts
    .sort_values("windows", ascending=False)
    .head(10)
    .to_string(index=False)
)

count    198.000000
mean      56.308081
std        6.335064
min        0.000000
25%       54.000000
50%       56.000000
75%       58.000000
max       93.000000
Name: windows, dtype: float64

Smallest 10 groups:
participant  session  windows
       p004        2        0
       p010        1       48
       p008        1       49
       p032        1       49
       p095        1       49
       p023        1       50
       p075        1       50
       p001        1       50
       p094        1       50
       p086        1       50

Largest 10 groups:
participant  session  windows
       p004        1       93
       p073        2       73
       p049        2       68
       p005        2       68
       p045        2       67
       p049        1       67
       p065        2       67
       p012        2       66
       p054        1       66
       p092        2       65


In [37]:
total_available = window_counts["windows"].sum()

target = 10000
selection_ratio = target / total_available

print("Available windows:", total_available)
print("Target windows:", target)
print("Selection ratio:", selection_ratio)

window_counts["target_float"] = (
    window_counts["windows"] * selection_ratio
)

print("\nTarget windows per group:")
print(
    window_counts[
        ["participant", "session", "windows", "target_float"]
    ].head(10).to_string(index=False)
)

Available windows: 11149
Target windows: 10000
Selection ratio: 0.896941429724639

Target windows per group:
participant  session  windows  target_float
       p001        1       50     44.847071
       p001        2       54     48.434837
       p002        1       54     48.434837
       p002        2       56     50.228720
       p003        1       51     45.744013
       p003        2       54     48.434837
       p004        1       93     83.415553
       p004        2        0      0.000000
       p005        1       65     58.301193
       p005        2       68     60.992017


In [38]:
import numpy as np

window_counts["target_windows"] = np.floor(
    window_counts["target_float"]
).astype(int)

print("Total after flooring:",
      window_counts["target_windows"].sum())

print("\nWindows still needed:",
      10000 - window_counts["target_windows"].sum())

Total after flooring: 9913

Windows still needed: 87


In [39]:
# Fractional part of each group's ideal allocation
window_counts["fraction"] = (
    window_counts["target_float"]
    - window_counts["target_windows"]
)

# Give one extra window to the 87 groups
# with the largest fractional parts
remaining = 10000 - window_counts["target_windows"].sum()

top_groups = window_counts["fraction"].nlargest(remaining).index

window_counts.loc[top_groups, "target_windows"] += 1

print("Final allocated windows:",
      window_counts["target_windows"].sum())

print("\nAllocation statistics:")
print(window_counts["target_windows"].describe())

print("\nGroups receiving an extra window:",
      (window_counts["fraction"].rank(method="first", ascending=False) <= remaining).sum())

Final allocated windows: 10000

Allocation statistics:
count    198.000000
mean      50.505051
std        5.674547
min        0.000000
25%       48.000000
50%       50.000000
75%       52.000000
max       83.000000
Name: target_windows, dtype: float64

Groups receiving an extra window: 87


In [40]:
windows = []

for (participant, session), group in clean_df.groupby(
    ["participant", "session"], sort=False
):
    group = group.reset_index(drop=True)

    # Number of windows allocated to this participant-session
    n_windows = int(
        window_counts.loc[
            (window_counts["participant"] == participant) &
            (window_counts["session"] == session),
            "target_windows"
        ].iloc[0]
    )

    # Create consecutive non-overlapping 50-event windows
    for window_id in range(n_windows):
        start = window_id * 50
        end = start + 50

        window = group.iloc[start:end].copy()

        # Safety check
        if len(window) == 50:
            windows.append({
                "participant": participant,
                "session": session,
                "window_id": window_id,
                "data": window
            })

print("Total windows created:", len(windows))

Total windows created: 10000


In [41]:
window_lengths = [len(w["data"]) for w in windows]

print("Number of windows:", len(window_lengths))
print("Minimum window length:", min(window_lengths))
print("Maximum window length:", max(window_lengths))
print("All windows have 50 events:", all(length == 50 for length in window_lengths))

Number of windows: 10000
Minimum window length: 50
Maximum window length: 50
All windows have 50 events: True


In [42]:
# Check that every window belongs to exactly one participant-session group
window_groups = [
    (w["participant"], w["session"])
    for w in windows
]

print("Unique participant-session groups in windows:",
      len(set(window_groups)))

print("Original participant-session groups:",
      len(window_counts))

# Count windows per participant-session
created_counts = pd.Series(window_groups).value_counts().sort_index()

print("\nFirst 10 participant-session window counts:")
print(created_counts.head(10))

print("\nTotal windows:", created_counts.sum())

Unique participant-session groups in windows: 197
Original participant-session groups: 198

First 10 participant-session window counts:
(p001, 1)    45
(p001, 2)    49
(p002, 1)    49
(p002, 2)    50
(p003, 1)    46
(p003, 2)    48
(p004, 1)    83
(p005, 1)    58
(p005, 2)    61
(p006, 1)    49
Name: count, dtype: int64

Total windows: 10000


In [43]:
# Check that window IDs are unique within each participant-session
window_ids = pd.DataFrame([
    {
        "participant": w["participant"],
        "session": w["session"],
        "window_id": w["window_id"]
    }
    for w in windows
])

duplicates = window_ids.duplicated(
    subset=["participant", "session", "window_id"]
).sum()

print("Duplicate window IDs:", duplicates)
print("Total windows:", len(window_ids))
print("Unique windows:", len(window_ids.drop_duplicates()))

Duplicate window IDs: 0
Total windows: 10000
Unique windows: 10000


In [44]:
# Check that window IDs are unique within each participant-session
window_ids = pd.DataFrame([
    {
        "participant": w["participant"],
        "session": w["session"],
        "window_id": w["window_id"]
    }
    for w in windows
])

duplicates = window_ids.duplicated(
    subset=["participant", "session", "window_id"]
).sum()

print("Duplicate window IDs:", duplicates)
print("Total windows:", len(window_ids))
print("Unique windows:", len(window_ids.drop_duplicates()))

Duplicate window IDs: 0
Total windows: 10000
Unique windows: 10000


In [45]:
# Columns available for feature extraction

timing_cols = [
    "DU.key1.key1",
    "DD.key1.key2",
    "DU.key1.key2",
    "UD.key1.key2",
    "UU.key1.key2"
]

for col in timing_cols:
    print(col)

DU.key1.key1
DD.key1.key2
DU.key1.key2
UD.key1.key2
UU.key1.key2


In [46]:
clean_df.columns = clean_df.columns.str.strip()

In [47]:
print(clean_df.columns.tolist())

['participant', 'session', 'key1', 'key2', 'DU.key1.key1', 'DD.key1.key2', 'DU.key1.key2', 'UD.key1.key2', 'UU.key1.key2']


In [48]:
timing_cols = [
    "DU.key1.key1",
    "DD.key1.key2",
    "DU.key1.key2",
    "UD.key1.key2",
    "UU.key1.key2"
]

In [49]:
clean_df[timing_cols].describe()

,DD.key1.key2,DU.key1.key2,UD.key1.key2,UU.key1.key2
count,562385.000000,562385.000000,562385.000000,562385.000000
mean,0.291818,0.495465,0.088476,0.292115
std,5.968022,5.490270,5.479273,4.635057
min,-568.083000,-7.935000,-568.162000,-10.782000
25%,0.120000,0.216000,0.019000,0.118000
50%,0.174000,0.272000,0.081000,0.176000
75%,0.272000,0.379000,0.173000,0.272000
max,1712.727000,1713.061000,1712.648000,1712.982000


In [50]:
print("clean_df exists:", "clean_df" in globals())
print("windows exists:", "windows" in globals())

clean_df exists: True
windows exists: True


In [51]:
import pandas as pd
import numpy as np

df = pd.read_csv("../data/raw/free-text.csv")
print(df.shape)

(562583, 10)


C:\Users\DELL\AppData\Local\Temp\ipykernel_23820\3653063053.py:4: DtypeWarning: Columns (4: DU.key1.key1) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv("../data/raw/free-text.csv")


In [52]:
df.columns = df.columns.str.strip()

print(df.columns.tolist())

['participant', 'session', 'key1', 'key2', 'DU.key1.key1', 'DD.key1.key2', 'DU.key1.key2', 'UD.key1.key2', 'UU.key1.key2', 'Unnamed: 9']


In [53]:
clean_df = df.dropna(
    subset=[
        "key2",
        "DD.key1.key2",
        "DU.key1.key2",
        "UD.key1.key2",
        "UU.key1.key2"
    ]
).copy()

clean_df = clean_df.drop(columns=["Unnamed: 9"])

print("Cleaned shape:", clean_df.shape)
print("Missing values:", clean_df.isna().sum().sum())

Cleaned shape: (562385, 9)
Missing values: 0


In [54]:
group_sizes = (
    clean_df
    .groupby(["participant", "session"])
    .size()
    .reset_index(name="rows")
)

group_sizes["available_windows"] = group_sizes["rows"] // 50

print("Total available 50-event windows:",
      group_sizes["available_windows"].sum())

print("\nParticipant-session groups:",
      len(group_sizes))

print("\nGroups with 0 windows:",
      (group_sizes["available_windows"] == 0).sum())

Total available 50-event windows: 11149

Participant-session groups: 198

Groups with 0 windows: 1


In [55]:
target_windows = 10000

group_sizes["target_float"] = (
    group_sizes["available_windows"]
    * target_windows
    / group_sizes["available_windows"].sum()
)

group_sizes["target_windows"] = (
    group_sizes["target_float"].astype(int)
)

remaining = (
    target_windows
    - group_sizes["target_windows"].sum()
)

print("Windows after flooring:", group_sizes["target_windows"].sum())
print("Remaining windows to allocate:", remaining)

Windows after flooring: 9913
Remaining windows to allocate: 87


In [56]:
group_sizes["fraction"] = (
    group_sizes["target_float"]
    - group_sizes["target_windows"]
)

extra_indices = (
    group_sizes["fraction"]
    .nlargest(remaining)
    .index
)

group_sizes.loc[extra_indices, "target_windows"] += 1

print("Final target windows:",
      group_sizes["target_windows"].sum())

Final target windows: 10000


In [57]:
windows = []

for (participant, session), group in clean_df.groupby(
    ["participant", "session"], sort=False
):
    
    group = group.reset_index(drop=True)

    n_windows = int(
        group_sizes.loc[
            (group_sizes["participant"] == participant) &
            (group_sizes["session"] == session),
            "target_windows"
        ].iloc[0]
    )

    for window_id in range(n_windows):
        
        start = window_id * 50
        end = start + 50
        
        window = group.iloc[start:end].copy()

        if len(window) == 50:
            windows.append({
                "participant": participant,
                "session": session,
                "window_id": window_id,
                "data": window
            })

print("Total windows created:", len(windows))

Total windows created: 10000


In [58]:
timing_cols = [
    "DU.key1.key1",
    "DD.key1.key2",
    "DU.key1.key2",
    "UD.key1.key2",
    "UU.key1.key2"
]

In [59]:
print(window[timing_cols].dtypes)

DU.key1.key1     object
DD.key1.key2    float64
DU.key1.key2    float64
UD.key1.key2    float64
UU.key1.key2    float64
dtype: object


In [60]:
clean_df["DU.key1.key1"] = pd.to_numeric(
    clean_df["DU.key1.key1"],
    errors="coerce"
)

In [61]:
print(clean_df["DU.key1.key1"].dtype)
print(clean_df["DU.key1.key1"].isna().sum())

float64
13


In [62]:
clean_df = clean_df.dropna(
    subset=["DU.key1.key1"]
).copy()

print("Shape after removing invalid DU values:", clean_df.shape)
print("Missing values:", clean_df.isna().sum().sum())

Shape after removing invalid DU values: (562372, 9)
Missing values: 0


In [63]:
group_sizes = (
    clean_df
    .groupby(["participant", "session"])
    .size()
    .reset_index(name="rows")
)

group_sizes["available_windows"] = group_sizes["rows"] // 50

print("Available 50-event windows:",
      group_sizes["available_windows"].sum())

Available 50-event windows: 11149


In [64]:
target_windows = 10000

group_sizes["target_float"] = (
    group_sizes["available_windows"]
    * target_windows
    / group_sizes["available_windows"].sum()
)

group_sizes["target_windows"] = (
    group_sizes["target_float"].astype(int)
)

remaining = (
    target_windows
    - group_sizes["target_windows"].sum()
)

print("After flooring:", group_sizes["target_windows"].sum())
print("Remaining:", remaining)

After flooring: 9913
Remaining: 87


In [65]:
group_sizes["fraction"] = (
    group_sizes["target_float"]
    - group_sizes["target_windows"]
)

extra_indices = (
    group_sizes["fraction"]
    .nlargest(remaining)
    .index
)

group_sizes.loc[extra_indices, "target_windows"] += 1

print("Final allocated windows:",
      group_sizes["target_windows"].sum())

Final allocated windows: 10000


In [66]:
windows = []

for (participant, session), group in clean_df.groupby(
    ["participant", "session"], sort=False
):
    
    group = group.reset_index(drop=True)

    n_windows = int(
        group_sizes.loc[
            (group_sizes["participant"] == participant) &
            (group_sizes["session"] == session),
            "target_windows"
        ].iloc[0]
    )

    for window_id in range(n_windows):
        
        start = window_id * 50
        end = start + 50
        
        window = group.iloc[start:end].copy()

        if len(window) == 50:
            windows.append({
                "participant": participant,
                "session": session,
                "window_id": window_id,
                "data": window
            })

print("Total windows created:", len(windows))

Total windows created: 10000


In [67]:
timing_cols = [
    "DU.key1.key1",
    "DD.key1.key2",
    "DU.key1.key2",
    "UD.key1.key2",
    "UU.key1.key2"
]

window = windows[0]["data"]

features = {}

for col in timing_cols:
    features[f"{col}_mean"] = window[col].mean()
    features[f"{col}_std"] = window[col].std()
    features[f"{col}_median"] = window[col].median()
    features[f"{col}_min"] = window[col].min()
    features[f"{col}_max"] = window[col].max()

features

{'DU.key1.key1_mean': np.float64(0.1209),
 'DU.key1.key1_std': np.float64(0.12603097286956477),
 'DU.key1.key1_median': np.float64(0.095),
 'DU.key1.key1_min': np.float64(0.056),
 'DU.key1.key1_max': np.float64(0.962),
 'DD.key1.key2_mean': np.float64(0.3116),
 'DD.key1.key2_std': np.float64(0.38387434168498635),
 'DD.key1.key2_median': np.float64(0.1835),
 'DD.key1.key2_min': np.float64(-0.796),
 'DD.key1.key2_max': np.float64(1.537),
 'DU.key1.key2_mean': np.float64(0.43190000000000006),
 'DU.key1.key2_std': np.float64(0.34815104920310735),
 'DU.key1.key2_median': np.float64(0.3045),
 'DU.key1.key2_min': np.float64(0.166),
 'DU.key1.key2_max': np.float64(1.631),
 'UD.key1.key2_mean': np.float64(0.1907),
 'UD.key1.key2_std': np.float64(0.3604236311308555),
 'UD.key1.key2_median': np.float64(0.1115),
 'UD.key1.key2_min': np.float64(-0.946),
 'UD.key1.key2_max': np.float64(1.413),
 'UU.key1.key2_mean': np.float64(0.311),
 'UU.key1.key2_std': np.float64(0.32133795548091276),
 'UU.key1.ke

In [68]:
print(window.shape)
print(window[timing_cols].dtypes)

(50, 9)
DU.key1.key1    float64
DD.key1.key2    float64
DU.key1.key2    float64
UD.key1.key2    float64
UU.key1.key2    float64
dtype: object


In [69]:
feature_rows = []

for w in windows:
    window = w["data"]

    features = {
        "participant": w["participant"],
        "session": w["session"],
        "window_id": w["window_id"]
    }

    for col in timing_cols:
        features[f"{col}_mean"] = window[col].mean()
        features[f"{col}_std"] = window[col].std()
        features[f"{col}_median"] = window[col].median()
        features[f"{col}_min"] = window[col].min()
        features[f"{col}_max"] = window[col].max()

    feature_rows.append(features)

keyboard_features = pd.DataFrame(feature_rows)

print("Shape:", keyboard_features.shape)
print("Rows:", len(keyboard_features))
print("Columns:", len(keyboard_features.columns))

Shape: (10000, 28)
Rows: 10000
Columns: 28


In [70]:
print("Missing values:",
      keyboard_features.isna().sum().sum())

Missing values: 0


In [71]:
keyboard_features.head()

,participant,session,window_id,DU.key1.key1_mean,DU.key1.key1_std,DU.key1.key1_median,DU.key1.key1_min,DU.key1.key1_max,DD.key1.key2_mean,DD.key1.key2_std,...,UD.key1.key2_mean,UD.key1.key2_std,UD.key1.key2_median,UD.key1.key2_min,UD.key1.key2_max,UU.key1.key2_mean,UU.key1.key2_std,UU.key1.key2_median,UU.key1.key2_min,UU.key1.key2_max
0,p001,1,0,0.12090,0.126031,0.0950,0.056,0.962,0.31160,0.383874,...,0.19070,0.360424,0.1115,-0.946,1.413,0.31100,0.321338,0.2240,0.016,1.507
1,p001,1,1,0.10006,0.035225,0.0865,0.055,0.184,0.69992,1.747901,...,0.59986,1.751623,0.1380,-0.048,12.071,0.69920,1.752654,0.2450,0.025,12.182
2,p001,1,2,0.10054,0.031475,0.0945,0.059,0.164,0.36300,0.482233,...,0.26246,0.482831,0.1060,-0.041,2.954,0.36316,0.476304,0.2115,0.040,3.032
3,p001,1,3,0.10018,0.041348,0.0895,0.051,0.304,0.28254,0.337433,...,0.18236,0.341688,0.0885,-0.288,1.809,0.28342,0.340618,0.1805,0.001,1.888
4,p001,1,4,0.15258,0.248436,0.1030,0.054,1.595,0.65990,1.732429,...,0.50732,1.665135,0.0960,-0.479,11.472,0.66016,1.861339,0.1920,0.040,13.067


In [72]:
print("Unique key1:", clean_df["key1"].nunique())
print("Unique key2:", clean_df["key2"].nunique())

print("\nMost common key1:")
print(clean_df["key1"].value_counts().head(15))

print("\nMost common key2:")
print(clean_df["key2"].value_counts().head(15))

Unique key1: 132
Unique key2: 132

Most common key1:
key1
Space        93868
e            47094
t            39266
Backspace    37449
o            36161
a            28022
i            26372
n            25326
s            23949
r            22179
h            21393
l            18962
u            15275
d            13657
m            10894
Name: count, dtype: int64

Most common key2:
key2
Space        93870
e            47180
t            39267
Backspace    37467
o            36160
a            28022
i            26372
n            25326
s            24031
r            22184
h            21393
l            18962
u            15275
d            13658
m            10894
Name: count, dtype: int64


In [73]:
window = windows[0]["data"]

key_features = {
    "unique_key1_count": window["key1"].nunique(),
    "unique_key2_count": window["key2"].nunique(),
    "unique_transition_count": window[["key1", "key2"]].drop_duplicates().shape[0]
}

key_features

{'unique_key1_count': 20,
 'unique_key2_count': 19,
 'unique_transition_count': 38}

In [74]:
window = windows[0]["data"]

same_key1 = (
    window["key1"].iloc[1:].values
    == window["key1"].iloc[:-1].values
)

repetition_rate = same_key1.mean()

print("Consecutive key1 repetition rate:", repetition_rate)

Consecutive key1 repetition rate: 0.0


In [75]:
repetition_rates = []

for w in windows[:100]:
    window = w["data"]

    same_key1 = (
        window["key1"].iloc[1:].values
        == window["key1"].iloc[:-1].values
    )

    repetition_rates.append(same_key1.mean())

print("Minimum:", min(repetition_rates))
print("Maximum:", max(repetition_rates))
print("Mean:", np.mean(repetition_rates))
print("Windows with repetition:", sum(r > 0 for r in repetition_rates))

Minimum: 0.0
Maximum: 0.32653061224489793
Mean: 0.037142857142857144
Windows with repetition: 68


In [76]:
window = windows[0]["data"]

backspace_count = (window["key1"] == "Backspace").sum()
backspace_rate = backspace_count / len(window)

print("Backspace count:", backspace_count)
print("Backspace rate:", backspace_rate)

Backspace count: 0
Backspace rate: 0.0


In [77]:
backspace_rates = []

for w in windows[:100]:
    window = w["data"]

    backspace_count = (window["key1"] == "Backspace").sum()
    backspace_rate = backspace_count / len(window)

    backspace_rates.append(backspace_rate)

print("Minimum:", min(backspace_rates))
print("Maximum:", max(backspace_rates))
print("Mean:", np.mean(backspace_rates))
print("Windows with Backspace:", sum(r > 0 for r in backspace_rates))

Minimum: 0.0
Maximum: 0.3
Mean: 0.035
Windows with Backspace: 49


In [78]:
for w in windows:
    window = w["data"]

    # Key diversity
    w["unique_key1_count"] = window["key1"].nunique()
    w["unique_key2_count"] = window["key2"].nunique()

    transitions = (
        window["key1"].astype(str) + " -> " +
        window["key2"].astype(str)
    )
    w["unique_transition_count"] = transitions.nunique()

    # Consecutive repetition
    same_key1 = (
        window["key1"].iloc[1:].values
        == window["key1"].iloc[:-1].values
    )
    w["consecutive_key1_repetition_rate"] = same_key1.mean()

    # Backspace behavior
    w["backspace_rate"] = (
        (window["key1"] == "Backspace").sum() / len(window)
    )

print("Features added successfully.")

Features added successfully.


In [79]:
print({
    "unique_key1_count": windows[0]["unique_key1_count"],
    "unique_key2_count": windows[0]["unique_key2_count"],
    "unique_transition_count": windows[0]["unique_transition_count"],
    "consecutive_key1_repetition_rate": windows[0]["consecutive_key1_repetition_rate"],
    "backspace_rate": windows[0]["backspace_rate"]
})

{'unique_key1_count': 20, 'unique_key2_count': 19, 'unique_transition_count': 38, 'consecutive_key1_repetition_rate': np.float64(0.0), 'backspace_rate': np.float64(0.0)}


In [80]:
keyboard_features = []

for w in windows:
    window = w["data"]

    row = {
        "participant": w["participant"],
        "session": w["session"],
        "window_id": w["window_id"]
    }

    # Timing features
    for col in timing_cols:
        row[f"{col}_mean"] = window[col].mean()
        row[f"{col}_std"] = window[col].std()
        row[f"{col}_median"] = window[col].median()
        row[f"{col}_min"] = window[col].min()
        row[f"{col}_max"] = window[col].max()

    # Key diversity
    row["unique_key1_count"] = window["key1"].nunique()
    row["unique_key2_count"] = window["key2"].nunique()

    transitions = (
        window["key1"].astype(str) + " -> " +
        window["key2"].astype(str)
    )
    row["unique_transition_count"] = transitions.nunique()

    # Consecutive repetition
    same_key1 = (
        window["key1"].iloc[1:].values
        == window["key1"].iloc[:-1].values
    )
    row["consecutive_key1_repetition_rate"] = same_key1.mean()

    # Backspace behavior
    row["backspace_rate"] = (
        (window["key1"] == "Backspace").sum() / len(window)
    )

    keyboard_features.append(row)

keyboard_features = pd.DataFrame(keyboard_features)

print("Shape:", keyboard_features.shape)
print("Missing values:", keyboard_features.isna().sum().sum())

Shape: (10000, 33)
Missing values: 0


In [81]:
print("Shape:", keyboard_features.shape)

print("\nDuplicate window IDs:",
      keyboard_features.duplicated(
          subset=["participant", "session", "window_id"]
      ).sum())

print("\nMissing values:",
      keyboard_features.isna().sum().sum())

print("\nFeature columns:")
print(keyboard_features.columns.tolist())

Shape: (10000, 33)

Duplicate window IDs: 0

Missing values: 0

Feature columns:
['participant', 'session', 'window_id', 'DU.key1.key1_mean', 'DU.key1.key1_std', 'DU.key1.key1_median', 'DU.key1.key1_min', 'DU.key1.key1_max', 'DD.key1.key2_mean', 'DD.key1.key2_std', 'DD.key1.key2_median', 'DD.key1.key2_min', 'DD.key1.key2_max', 'DU.key1.key2_mean', 'DU.key1.key2_std', 'DU.key1.key2_median', 'DU.key1.key2_min', 'DU.key1.key2_max', 'UD.key1.key2_mean', 'UD.key1.key2_std', 'UD.key1.key2_median', 'UD.key1.key2_min', 'UD.key1.key2_max', 'UU.key1.key2_mean', 'UU.key1.key2_std', 'UU.key1.key2_median', 'UU.key1.key2_min', 'UU.key1.key2_max', 'unique_key1_count', 'unique_key2_count', 'unique_transition_count', 'consecutive_key1_repetition_rate', 'backspace_rate']


In [86]:
keyboard_features.to_csv(
    "../data/processed/keyboard_features.csv",
    index=False
)

print("keyboard_features.csv saved successfully.")

keyboard_features.csv saved successfully.
